### DISCUSSION TITLE
Client behavior with degraded cluster seems to be wrong

### DISCUSSION THREAD

**Discussion by krivulcik** [ravendb/ravendb#14454] [Q&A]
We observe client behavior that doesn't match our expectations when a database is in a degraded state.

While one of the nodes (node A) was being rebuilt (receiving replicated documents), the client applications kept trying to connect to that node and even managed to receive outdated data from the degraded node.

Here's what we have done to prevent this behavior (incrementally):

 - The first attempt was to let the node recover without any intervention. This is done automatically.
   - Our applications worked correctly for a period of time. After some time, the applications started to connect to the degraded node and fail (due to missing data, etc.).
 - We have paused the indexing on the degraded node to a) lower the load on the server and b) prevent the node from going from Rehab to Member automatically (we wanted to oversee the transition).
   - Even in this state, the applications started to connect to the degraded node and fail (due to missing data, incomplete indexes, etc.).
 - After that, we have blocked incoming traffic to ports 80, 443, and 38888 from all sources except the other two nodes to prevent the applications from using the node.
   - In this state, the database group kept deciding that the database on node A was healthy (see the screenshot in https://github.com/ravendb/ravendb/discussions/14448#discussion-4176316 ).
   - The applications started to connect to the degraded node and fail (due to timeouts).

It seems that even though the database should be considered in Rehab, this is not the case. Consequently, applications attempt to use the database on the degraded node and fail.

I believe that the node should be aware of the fact that the database is not ready to use, and this information should be consistent across the nodes in the database group - is this the case?

What are situations when the database group state information can be inconsistent between the individual nodes? Can it be influenced by the cluster not having a leader (solved in https://github.com/ravendb/ravendb/discussions/14446 ).

The above has happened even after the cluster leader was successfully elected and took office. After solving the issue in discussion 14446, we have started node A in the above configuration - indexing paused, firewall blocking all traffic except from the other cluster nodes. Overnight, the applications began connecting to the node and failing. Also, the documents seem to be incompletely replicated (as described in discussion 14448, there are documents with outdated data and missing documents).

We are currently provisioning a new cluster node and will add it to the cluster and the database groups.

Do you have any suggestions to prevent a similar situation from happening with the new node?

**Comment by garayx** [comment]
Hi

The readded node should be in a `rehab` state in the database group of all the databases that this node has,
so the client will fetch the updates topology and connect to `healthy` node (the ones above the readded node in topology group), but if you don't have topology updates available, then the client will always connect to 1st node you passed to the documentstore, even if the node is in rehab state. Can you check if you have topology updates disabled ? if it is then such situation can happen.

**Reply by krivulcik** [reply]
We have also removed the readded node from the configuration (so the Urls field contains only nodes B and C). Topology updates are not disabled. The topology should be updated normally.

We have also observed the following behavior:
 - After starting, node A was in rehab state and was replicating the documents.
 - After some time, the client web application began using node A (and failing). This somehow stuck - the application kept failing until manual restart, after which it retrieved the topology and started using the healthy nodes (we brought node A down then).

It seems that during the operation, something caused the application to try connecting to node A. This might have been a timeout or an intermittent network failure or a failed database request, or whatever. It then tried to use node A, but node A was unreachable (blocked at the firewall to prevent exactly this - the application using the database and potentially serving incorrect data or writing data to the node in rehab). But the client application wouldn't recognize that node A was unreachable and wouldn't start using the healthy nodes.

We have no clue why this behavior could arise. From what we understand, the client should try the next node in the database group, right?

I'm not sure what was the status of the database group on the individual nodes during this incident, but should the client retrieve a new topology when a connection attempt to the preferred node times out?

**Reply by garayx** [reply]
If topology updates are enabled then node's A url will be fetched with the topology update, even if its not stated in urls passed to store.

The cluster itself was stable at this point ? Was there a leader ?

**Reply by krivulcik** [reply]
The cluster was as stable as it gets - the voting was occurring intermittently, but node C was the leader most of the time.

I don't have hard data about the voting rate and don't know what rate is considered normal. For comparison, we have the cluster in a degraded state right now, and the new node D just finished receiving documents around 20 minutes ago. In around 100 minutes (80 minutes receiving data, 20 minutes idle), the term increased by 90, which is an average voting rate of around 1/minute. I suspect that it was higher during the replication and nearly zero after finishing, so the voting rate during replication was probably higher than 1/minute.

However, with the new node D (more powerful than nodes A, B, and C), node A removed from the cluster, and all the system databases synchronized, the client behavior seems to have been okay during this last replication attempt.

I'm not sure about the root cause, but it was likely a combination of high load due to replication + indexing, unstable cluster without a leader, and intermittent timeouts causing the node in rehab to receive traffic and changed data.

**Reply by ayende** [reply]
We had a bug where a database would move to rehab on an unneeded basis, that can cause us to change database topology too often.
What exact build are you running here?

**Reply by krivulcik** [reply]
We are running 5.3.103 since shortly after it was released.

**Reply by ayende** [reply]
So that is after that.
Note that 1 vote / minute is pretty high. We usually see hours or days if there are no errors.

Right now, do you have any instability? Can you share the cluster observer log?

![image](https://user-images.githubusercontent.com/116915/176478655-1d29c714-4af3-4077-a892-11d3226f2871.png)


![image](https://user-images.githubusercontent.com/116915/176478606-2204708e-4052-4386-83b6-b071375beb72.png)

**Reply by krivulcik** [reply]
I'll try to capture some logs. However, we encounter a situation I believe is not nominal: When I try to do any cluster operation (reassigning cores among nodes, promoting/demoting, stepping down, etc.) on a node that is not currently the leader, I receive an error:

This server requires client certificate for authentication, but none was provided by the client. Did you forget to install the certificate?
Your certificate store may be cached by the browser. Create a new private browsing tab, which will not cache any certificates. (Ctrl+Shift+N in Chrome, Ctrl+Shift+P in Firefox) Operator/ClusterAdmin access is required but not given to this certificate

I believe that when I submit the request to a non-leader node, the node redirects/forwards the request to the leader. However, the forwarded request is not authenticated correctly towards the leader node.

The same issue happens with collecting the Cluster Observer Log. Since it is available only on the leader node, I receive the same error. When the cluster topology changes, the log doesn't collect any entries on the leader node and needs to be opened on the new leader. This makes collecting it difficult to collect a meaningful amount of logs.

Do you have a recommendation on how to set up the nodes to allow them to redirect the requests to the leader correctly?

**Reply by krivulcik** [reply]
So far (a few minutes), the Cluster Observer Log behaves as follows:

 - I can open it only on the leader node due to the above issue.
 - The only entries in the log are messages related to database group topology changes like this: The database '...' on D is not ready to be promoted, because Index '...' is in state 'Normal' and not up-to-date (prev: 21857712, current: 2055045).
 - After a while, it displays the following message at the bottom and stops logging: A term has changed. Your results are stale. [Refresh.]
 - Refreshing it using the button Refresh doesn't work (yields the message with failed authentication).
 - When I open the log on the newly elected leader, it starts empty and is getting filled with the database group topology change messages.
 - I have not noticed a message related to cluster topology change yet.

**Reply by krivulcik** [reply]
The term has increased from 335,219 at 18:35 today to 335,248 at 19:08 (by 29 in 33 minutes) and then to 335,251 at 19:16 (by 3 in 8 minutes).
Would the Information-level admin log be useful to see what's going on?

**Reply by ayende** [reply]
That *should not* be possible. The " server requires client certificate for authentication"

That _usually_ indicates that you are in an environment where the HTTPS traffic is intercepted.
In that case, that would _utterly break_ many assumptions that we make.
What happens when you try `/admin/debug/node/ping` endpoint from each of the nodes?

Effectively, that means that there is _no communication_ in the cluster, and the nodes are operating independently.
Note that replication, on the other hand, isn't using HTTPS, so isn't subject to this issue. But when we'll need to refresh the topology, we'll run into the same issue.

**Reply by krivulcik** [reply]
I sent the contents of the `/admin/debug/node/ping` endpoints to your email, @ayende.

The cluster works correctly most of the time, and we've had this issue since we first installed the cluster.

As far as I can tell, the communication within the cluster works okay, but I'll let you evaluate that based on the data.

I tried to create a debug package before, but it took a very long time - tens of minutes - before I aborted it. Should I wait for the package however long it takes to create it if I need to create it?

**Reply by ayende** [reply]
It should be < 1 minute to create it.
See my reply to your email
